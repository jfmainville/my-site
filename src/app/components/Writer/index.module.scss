@import "../../variables.module.scss";

.Writer {
  display: flex;
  flex-direction: column;
}

.WriterInput {
  display: flex;
  height: 2rem;
  width: 20rem;
  margin: 0.5rem 0 0.5rem 1rem;
}

.WriterContent {
  margin-left: 1rem;
  background-color: #ffffff;
  border: 1px $primary-color;
  min-height: 50px;
}

.WriterButtons {
  display: flex;
  flex-direction: row;
}

.Link {
  font-size: 1.17em;
  color: $text-color;
  padding: 0 1rem 0 1rem;
}
