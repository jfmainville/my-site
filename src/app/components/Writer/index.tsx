"use client";

import styles from "./index.module.scss";
import { useState, useCallback } from "react";
import { EditorContent, useEditor } from "@tiptap/react";
import StarterKit from "@tiptap/starter-kit";
import Image from "@tiptap/extension-image";
import { MY_SITE_URL } from "@/app/utils/constants";
import { useRouter } from "next/navigation";
import Button from "../../components/Button/index";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faBold,
  faItalic,
  faList,
  faListOl,
  faStrikethrough,
  faCode,
  faImage,
} from "@fortawesome/free-solid-svg-icons";

const Writer = ({ postData }: any) => {
  const router = useRouter();
  const [postTitle, setPostTitle] = useState(postData ? postData.title : "");
  const [postThumbnail, setPostThumbnail] = useState(
    postData ? postData.thumbnail : "",
  );
  const [postStatus, setPostStatus] = useState(
    postData ? postData.status : "DRAFT",
  );
  const [postContent, setPostContent] = useState(
    postData ? postData.content : "Enter text here...",
  );
  const [postCategory, setPostCategory] = useState(
    postData ? postData.category : "",
  );
  const editor = useEditor({
    editorProps: {
      attributes: {
        class: styles.WriterEditor,
      },
    },
    extensions: [StarterKit, Image],
    content: postContent,
    onBlur({ editor }) {
      setPostContent(editor.getHTML());
    },
    immediatelyRender: false, // Need to pass this attribute to prevent rehydration mismatch when using SSR
  });

  const handleOnPostSave = async (
    postTitle: String,
    postThumbnail: String,
    postContent: String,
    postCategory: String,
  ) => {
    try {
      if (!postData) {
        try {
          await fetch(`${MY_SITE_URL}/api/post`, {
            method: "POST",
            headers: {
              Accept: "application/json",
              "Content-Type": "application/json",
            },
            body: JSON.stringify({
              title: postTitle,
              thumbnail: postThumbnail,
              slug: postTitle.replace(" ", "-").toLowerCase(),
              content: postContent,
              status: "DRAFT",
              category: postCategory,
              // TODO: Need to use the current userId as reference instead of a hardcoded value
              userId: "seed-user-1",
            }),
          });
        } catch (e) {
          console.error("Failed to create the post in the database");
        }
      } else if (postData) {
        try {
          await fetch(`${MY_SITE_URL}/api/post`, {
            method: "PUT",
            headers: {
              Accept: "application/json",
              "Content-Type": "application/json",
            },
            body: JSON.stringify({
              id: postData.id,
              title: postTitle,
              thumbnail: postThumbnail,
              slug: postTitle.replace(" ", "-").toLowerCase(),
              content: postContent,
              status: postData.status,
              category: postCategory,
              // TODO: Need to use the current userId as reference instead of a hardcoded value
              userId: "seed-user-2",
            }),
          });
        } catch (e) {
          console.error("Failed to update the post in the database");
        }
      }

      // Navigate back to the admin page
      router.push("/admin/post");

      // Rehydrate the state after navigating back the admin page
      router.refresh();
    } catch (e) {
      console.error("Failed to save the post in the database", e);
    }
  };

  const handleOnPostDelete = async (postId: String) => {
    try {
      if (postId) {
        await fetch(`${MY_SITE_URL}/api/post`, {
          method: "DELETE",
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            id: postId,
          }),
        });
      }

      // Navigate back to the admin page
      router.push("/admin/post");

      // Rehydrate the state after navigating back the admin page
      router.refresh();
    } catch (e) {
      console.error("Failed to create the post in the database");
    }
  };

  const addImage = useCallback(() => {
    const url = window.prompt("URL");

    if (url) {
      editor.chain().focus().setImage({ src: url }).run();
    }
  }, [editor]);
  if (!editor) {
    return null;
  }

  return (
    <div className={styles.Writer}>
      <input
        className={styles.WriterInput}
        value={postTitle}
        placeholder="Title"
        required={true}
        onChange={(event) => setPostTitle(event.target.value)}
      />
      <input
        className={styles.WriterInput}
        value={postCategory}
        placeholder="Category"
        required={true}
        onChange={(event) => setPostCategory(event.target.value)}
      />
      <input
        className={styles.WriterInput}
        value={postThumbnail}
        placeholder="Thumbnail"
        onChange={(event) => setPostThumbnail(event.target.value)}
      />
      <div className={styles.WriterEditorBar}>
        <button
          onClick={() => editor.chain().focus().toggleBold().run()}
          className={
            editor.isActive("bold")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faBold} width="40" height="40" />
        </button>
        <button
          onClick={() => editor.chain().focus().toggleItalic().run()}
          className={
            editor.isActive("italic")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faItalic} width="40" height="40" />
        </button>
        <button
          onClick={() => editor.chain().focus().toggleStrike().run()}
          className={
            editor.isActive("strike")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faStrikethrough} width="40" height="40" />
        </button>
        <button
          onClick={() =>
            editor.chain().focus().toggleHeading({ level: 1 }).run()
          }
          className={
            editor.isActive("heading", { level: 1 })
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
          style={{ width: "2.5rem" }}
        >
          H1
        </button>
        <button
          onClick={() =>
            editor.chain().focus().toggleHeading({ level: 2 }).run()
          }
          className={
            editor.isActive("heading", { level: 2 })
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
          style={{ width: "2.5rem" }}
        >
          H2
        </button>
        <button
          onClick={() =>
            editor.chain().focus().toggleHeading({ level: 3 }).run()
          }
          className={
            editor.isActive("heading", { level: 3 })
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
          style={{ width: "2.5rem" }}
        >
          H3
        </button>
        <button
          onClick={() => editor.chain().focus().toggleBulletList().run()}
          className={
            editor.isActive("bulletList")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faList} width="40" height="40" />
        </button>
        <button
          onClick={() => editor.chain().focus().toggleOrderedList().run()}
          className={
            editor.isActive("orderedList")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faListOl} width="40" height="40" />
        </button>
        <button
          onClick={() => editor.chain().focus().toggleCodeBlock().run()}
          className={
            editor.isActive("codeBlock")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faCode} width="40" height="40" />
        </button>
        <button
          onClick={addImage}
          className={
            editor.isActive("image")
              ? styles.WriterEditorButtonActive
              : styles.WriterEditorButton
          }
        >
          <FontAwesomeIcon icon={faImage} width="40" height="40" />
        </button>
      </div>
      <EditorContent data-testid="TipTap" editor={editor} />
      <div className={styles.WriterButtons}>
        <Button
          onButtonClickEvent={() =>
            handleOnPostSave(postTitle, postContent, postCategory)
          }
          buttonText={"Save"}
        />
        {postData ? (
          <Button
            onButtonClickEvent={() => handleOnPostDelete(postData.id)}
            buttonText={"Delete"}
            buttonColor="red"
          />
        ) : null}
      </div>
    </div>
  );
};

export default Writer;
