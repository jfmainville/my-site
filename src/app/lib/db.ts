"use server";

import { PrismaClient, User, Post, PostStatus } from "@prisma/client";
import { redirect } from "next/navigation";

export type UserSession = {
  user: {
    nickname: string;
    name: string;
    picture: string;
    updated_at: string;
    email: string;
    email_verified: boolean;
    sub: string;
    sid: string;
    accessToken: string;
    accessTokenScope: string;
    accessTokenExpiresAt: string;
    idToken: string;
    token_type: string;
  };
};

const prisma = new PrismaClient();

export async function getPosts() {
  let posts;

  posts = await prisma.post.findMany();

  return posts;
}

export async function getUniquePost(slug: String) {
  let posts;

  if (slug) {
    posts = await prisma.post.findUnique({
      where: {
        slug: slug as string,
      },
    });

    return posts;
  }
}

export async function createPost(
  postTitle: String,
  postCategory: String,
  postThumbnail: String,
  postStatus: String,
  postContent: String,
) {
  let post;

  post = {
    title: postTitle as string,
    thumbnail: postThumbnail as string | null,
    slug: postTitle.replace(" ", "-").toLowerCase() as string,
    content: postContent as string,
    status: (postStatus as PostStatus) || "DRAFT",
    category: postCategory as string,
    // TODO: Need to use the current userId as reference instead of a hardcoded value
    userId: "seed-user-1",
  };
  await prisma.post.create({
    data: post,
  });
  redirect("/admin/post");
}

export async function updatePost(
  postData: Post,
  postTitle: String,
  postThumbnail: String,
  postContent: String,
  postCategory: String,
  postStatus: String,
) {
  let post;

  post = {
    id: postData.id,
    title: postTitle as string,
    thumbnail: postThumbnail as string | null,
    slug: postTitle.replace(" ", "-").toLowerCase() as string,
    content: postContent as string,
    status: postStatus as PostStatus,
    category: postCategory as string,
    // TODO: Need to use the current userId as reference instead of a hardcoded value
    userId: "seed-user-2",
  };
  await prisma.post.update({
    where: {
      id: post.id,
    },
    data: post,
  });
  redirect("/admin/post");
}

export async function deletePost(postId: String) {
  await prisma.post.delete({
    where: {
      id: postId as string,
    },
  });
  redirect("/admin/post");
}

export async function handleUserSession(userSession: UserSession) {
  if (userSession && userSession.user) {
    const userData: User | null = await prisma.user.findUnique({
      where: {
        username: userSession.user?.nickname as string,
      },
    });

    if (!userData) {
      await prisma.user.create({
        data: {
          username: userSession.user.nickname as string,
          email: (userSession.user.email as string) || "",
          firstName: (userSession.user.name?.split(" ")[0] as string) || "",
          lastName: (userSession.user.name?.split(" ")[1] as string) || "",
        },
      });
    }
  }
}
