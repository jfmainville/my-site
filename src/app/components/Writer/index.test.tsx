import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Writer from "./index";
import { createPost, updatePost, deletePost } from "../../lib/db";

jest.mock("../../lib/db");

describe("<Writer/> Test Suite", () => {
  it("should contain the Writer required fields in the <Writer/> component", () => {
    render(<Writer />);
    const postTitle = screen.getByPlaceholderText("Title");
    const postCategory = screen.getByPlaceholderText("Category");
    const postThumbnail = screen.getByPlaceholderText("Thumbnail");
    const postStatus = screen.getByPlaceholderText("Status");
    const postContent = screen.getByTestId("TipTap");
    expect(postTitle).toBeInTheDocument();
    expect(postCategory).toBeInTheDocument();
    expect(postThumbnail).toBeInTheDocument();
    expect(postStatus).toBeInTheDocument();
    expect(postContent).toBeInTheDocument();
  });

  it("should fill the input fields if the postData props are being passed in the <Writer/> component", () => {
    const postData = {
      id: "12345",
      title: "Test Title",
      category: "Cybersecurity",
      thumbnail:
        "https://images.unsplash.com/photo-1725714834773-b5f278346915?q=80&w=2970&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D",
      status: "DRAFT",
      content: "<h1>Test Content</h1>",
    };
    render(<Writer postData={postData} />);
    const postTitle = screen.getByPlaceholderText("Title");
    const postCategory = screen.getByPlaceholderText("Category");
    const postThumbnail = screen.getByPlaceholderText("Thumbnail");
    const postStatus = screen.getByPlaceholderText("Status");
    expect(postTitle).toHaveValue("Test Title");
    expect(postCategory).toHaveValue("Cybersecurity");
    expect(postThumbnail).toHaveValue(
      "https://images.unsplash.com/photo-1725714834773-b5f278346915?q=80&w=2970&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D",
    );
    expect(postStatus).toHaveValue("DRAFT");
  });

  it("should show the delete button if the postData props are being passed in the <Writer/> component", () => {
    const postData = {
      id: "12345",
      title: "Test Title",
      category: "Cybersecurity",
      thumbnail:
        "https://images.unsplash.com/photo-1725714834773-b5f278346915?q=80&w=2970&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D",
      status: "DRAFT",
      content: "<h1>Test Content</h1>",
    };
    render(<Writer postData={postData} />);
    const postDeleteButton = screen.getByText("Delete");
    expect(postDeleteButton).toBeInTheDocument();
  });

  it("should not show the delete button if the postData props are not being passed in the <Writer/> component", () => {
    render(<Writer />);
    const postDeleteButton = screen.queryByText("Delete");
    expect(postDeleteButton).toBeNull();
  });

  it("should fill the input fields in the <Writer/> component", () => {
    render(<Writer />);
    const postTitle = screen.getByPlaceholderText("Title");
    const postCategory = screen.getByPlaceholderText("Category");
    const postThumbnail = screen.getByPlaceholderText("Thumbnail");
    const postStatus = screen.getByPlaceholderText("Status");

    fireEvent.change(postTitle, { target: { value: "Test Title" } });
    fireEvent.change(postCategory, { target: { value: "Cybersecurity" } });
    fireEvent.change(postThumbnail, { target: { value: "" } });
    fireEvent.change(postStatus, { target: { value: "DRAFT" } });

    expect(postTitle).toHaveValue("Test Title");
    expect(postCategory).toHaveValue("Cybersecurity");
    expect(postThumbnail).toHaveValue("");
    expect(postStatus).toHaveValue("DRAFT");
  });

  it("should create the post data in the <Writer/> component", () => {
    const userData = {
      id: "seed-user-1",
      firstName: "Tony",
      lastName: "Stark",
      username: "tstark",
      email: "tstark@test.com",
      createdAt: "2025-01-20T01:51:31.356Z",
      updatedAt: "2025-01-20T01:51:31.356Z",
    };

    render(<Writer userData={userData} />);
    const postTitle = screen.getByPlaceholderText("Title");
    const postCategory = screen.getByPlaceholderText("Category");
    const postThumbnail = screen.getByPlaceholderText("Thumbnail");
    const postStatus = screen.getByPlaceholderText("Status");
    const postCreateButton = screen.getByText("Create");

    fireEvent.change(postTitle, { target: { value: "Test Title" } });
    fireEvent.change(postCategory, { target: { value: "Cybersecurity" } });
    fireEvent.change(postThumbnail, { target: { value: "" } });
    fireEvent.change(postStatus, { target: { value: "DRAFT" } });

    fireEvent.click(postCreateButton);

    expect(createPost as jest.Mock).toHaveBeenCalledTimes(1);
    expect(createPost as jest.Mock).toHaveBeenCalledWith(
      "Test Title",
      "Cybersecurity",
      "",
      "DRAFT",
      "Enter text here...",
      {
        id: "seed-user-1",
        firstName: "Tony",
        lastName: "Stark",
        username: "tstark",
        email: "tstark@test.com",
        createdAt: "2025-01-20T01:51:31.356Z",
        updatedAt: "2025-01-20T01:51:31.356Z",
      },
    );
  });

  it("should update the post in the <Writer/> component", () => {
    const postData = {
      id: "12345",
      title: "Test Title",
      category: "Cybersecurity",
      thumbnail: "",
      status: "DRAFT",
      content: "<h1>Test Content</h1>",
    };
    const userData = {
      id: "seed-user-1",
      firstName: "Tony",
      lastName: "Stark",
      username: "tstark",
      email: "tstark@test.com",
      createdAt: "2025-01-20T01:51:31.356Z",
      updatedAt: "2025-01-20T01:51:31.356Z",
    };

    render(<Writer postData={postData} userData={userData} />);
    const postSaveButton = screen.getByText("Update");

    fireEvent.click(postSaveButton);

    expect(updatePost as jest.Mock).toHaveBeenCalledTimes(1);
    expect(updatePost as jest.Mock).toHaveBeenCalledWith(
      {
        id: "12345",
        title: "Test Title",
        category: "Cybersecurity",
        thumbnail: "",
        status: "DRAFT",
        content: "<h1>Test Content</h1>",
      },
      "Test Title",
      "",
      "<h1>Test Content</h1>",
      "Cybersecurity",
      "DRAFT",
      {
        id: "seed-user-1",
        firstName: "Tony",
        lastName: "Stark",
        username: "tstark",
        email: "tstark@test.com",
        createdAt: "2025-01-20T01:51:31.356Z",
        updatedAt: "2025-01-20T01:51:31.356Z",
      },
    );
  });

  it("should delete the post in the <Writer/> component", () => {
    const postData = {
      id: "12345",
      title: "Test Title",
      category: "Cybersecurity",
      thumbnail: "",
      status: "DRAFT",
      content: "<h1>Test Content</h1>",
    };
    render(<Writer postData={postData} />);
    const postDeleteButton = screen.getByText("Delete");

    fireEvent.click(postDeleteButton);

    expect(deletePost as jest.Mock).toHaveBeenCalledTimes(1);
    expect(deletePost as jest.Mock).toHaveBeenCalledWith("12345");
  });
});
