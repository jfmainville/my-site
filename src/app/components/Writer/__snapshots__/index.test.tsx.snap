// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Writer/> Test Suite should render the <Writer/> component unchanged 1`] = `
<div>
  <div
    class="Writer"
  >
    <input
      class="WriterInput"
      placeholder="Title"
      required=""
      value=""
    />
    <input
      class="WriterInput"
      placeholder="Category"
      required=""
      value=""
    />
    <input
      class="WriterInput"
      placeholder="Thumbnail"
      value=""
    />
    <input
      class="WriterInput"
      placeholder="Status"
      value="DRAFT"
    />
    <div
      class="WriterEditorBar"
    >
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bold "
          data-icon="bold"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 384 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 64C0 46.3 14.3 32 32 32l48 0 16 0 128 0c70.7 0 128 57.3 128 128c0 31.3-11.3 60.1-30 82.3c37.1 22.4 62 63.1 62 109.7c0 70.7-57.3 128-128 128L96 480l-16 0-48 0c-17.7 0-32-14.3-32-32s14.3-32 32-32l16 0 0-160L48 96 32 96C14.3 96 0 81.7 0 64zM224 224c35.3 0 64-28.7 64-64s-28.7-64-64-64L112 96l0 128 112 0zM112 288l0 128 144 0c35.3 0 64-28.7 64-64s-28.7-64-64-64l-32 0-112 0z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-italic "
          data-icon="italic"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 384 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M128 64c0-17.7 14.3-32 32-32l192 0c17.7 0 32 14.3 32 32s-14.3 32-32 32l-58.7 0L160 416l64 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 480c-17.7 0-32-14.3-32-32s14.3-32 32-32l58.7 0L224 96l-64 0c-17.7 0-32-14.3-32-32z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-strikethrough "
          data-icon="strikethrough"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 512 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M161.3 144c3.2-17.2 14-30.1 33.7-38.6c21.1-9 51.8-12.3 88.6-6.5c11.9 1.9 48.8 9.1 60.1 12c17.1 4.5 34.6-5.6 39.2-22.7s-5.6-34.6-22.7-39.2c-14.3-3.8-53.6-11.4-66.6-13.4c-44.7-7-88.3-4.2-123.7 10.9c-36.5 15.6-64.4 44.8-71.8 87.3c-.1 .6-.2 1.1-.2 1.7c-2.8 23.9 .5 45.6 10.1 64.6c4.5 9 10.2 16.9 16.7 23.9L32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l448 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-209.9 0-.4-.1-1.1-.3c-36-10.8-65.2-19.6-85.2-33.1c-9.3-6.3-15-12.6-18.2-19.1c-3.1-6.1-5.2-14.6-3.8-27.4zM348.9 337.2c2.7 6.5 4.4 15.8 1.9 30.1c-3 17.6-13.8 30.8-33.9 39.4c-21.1 9-51.7 12.3-88.5 6.5c-18-2.9-49.1-13.5-74.4-22.1c-5.6-1.9-11-3.7-15.9-5.4c-16.8-5.6-34.9 3.5-40.5 20.3s3.5 34.9 20.3 40.5c3.6 1.2 7.9 2.7 12.7 4.3c0 0 0 0 0 0s0 0 0 0c24.9 8.5 63.6 21.7 87.6 25.6c0 0 0 0 0 0l.2 0c44.7 7 88.3 4.2 123.7-10.9c36.5-15.6 64.4-44.8 71.8-87.3c3.6-21 2.7-40.4-3.1-58.1l-75.7 0c7 5.6 11.4 11.2 13.9 17.2z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
        style="width: 2.5rem;"
      >
        H1
      </button>
      <button
        class="WriterEditorButton"
        style="width: 2.5rem;"
      >
        H2
      </button>
      <button
        class="WriterEditorButton"
        style="width: 2.5rem;"
      >
        H3
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list "
          data-icon="list"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 512 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M40 48C26.7 48 16 58.7 16 72l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24L40 48zM192 64c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L192 64zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-288 0zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-288 0zM16 232l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24l-48 0c-13.3 0-24 10.7-24 24zM40 368c-13.3 0-24 10.7-24 24l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24l-48 0z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list-ol "
          data-icon="list-ol"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 512 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M24 56c0-13.3 10.7-24 24-24l32 0c13.3 0 24 10.7 24 24l0 120 16 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-80 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l16 0 0-96-8 0C34.7 80 24 69.3 24 56zM86.7 341.2c-6.5-7.4-18.3-6.9-24 1.2L51.5 357.9c-7.7 10.8-22.7 13.3-33.5 5.6s-13.3-22.7-5.6-33.5l11.1-15.6c23.7-33.2 72.3-35.6 99.2-4.9c21.3 24.4 20.8 60.9-1.1 84.7L86.8 432l33.2 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-88 0c-9.5 0-18.2-5.6-22-14.4s-2.1-18.9 4.3-25.9l72-78c5.3-5.8 5.4-14.6 .3-20.5zM224 64l256 0c17.7 0 32 14.3 32 32s-14.3 32-32 32l-256 0c-17.7 0-32-14.3-32-32s14.3-32 32-32zm0 160l256 0c17.7 0 32 14.3 32 32s-14.3 32-32 32l-256 0c-17.7 0-32-14.3-32-32s14.3-32 32-32zm0 160l256 0c17.7 0 32 14.3 32 32s-14.3 32-32 32l-256 0c-17.7 0-32-14.3-32-32s14.3-32 32-32z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-code "
          data-icon="code"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 640 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M392.8 1.2c-17-4.9-34.7 5-39.6 22l-128 448c-4.9 17 5 34.7 22 39.6s34.7-5 39.6-22l128-448c4.9-17-5-34.7-22-39.6zm80.6 120.1c-12.5 12.5-12.5 32.8 0 45.3L562.7 256l-89.4 89.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l112-112c12.5-12.5 12.5-32.8 0-45.3l-112-112c-12.5-12.5-32.8-12.5-45.3 0zm-306.7 0c-12.5-12.5-32.8-12.5-45.3 0l-112 112c-12.5 12.5-12.5 32.8 0 45.3l112 112c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256l89.4-89.4c12.5-12.5 12.5-32.8 0-45.3z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="WriterEditorButton"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-image "
          data-icon="image"
          data-prefix="fas"
          focusable="false"
          height="40"
          role="img"
          viewBox="0 0 512 512"
          width="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 96C0 60.7 28.7 32 64 32l384 0c35.3 0 64 28.7 64 64l0 320c0 35.3-28.7 64-64 64L64 480c-35.3 0-64-28.7-64-64L0 96zM323.8 202.5c-4.5-6.6-11.9-10.5-19.8-10.5s-15.4 3.9-19.8 10.5l-87 127.6L170.7 297c-4.6-5.7-11.5-9-18.7-9s-14.2 3.3-18.7 9l-64 80c-5.8 7.2-6.9 17.1-2.9 25.4s12.4 13.6 21.6 13.6l96 0 32 0 208 0c8.9 0 17.1-4.9 21.2-12.8s3.6-17.4-1.4-24.7l-120-176zM112 192a48 48 0 1 0 0-96 48 48 0 1 0 0 96z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      data-testid="TipTap"
    >
      <div
        class="tiptap ProseMirror WriterEditor"
        contenteditable="true"
        tabindex="0"
        translate="no"
      >
        <p>
          Enter text here...
        </p>
      </div>
    </div>
    <div
      class="WriterButtons"
    >
      <div
        class="Button"
      >
        <div>
          Create
        </div>
      </div>
    </div>
  </div>
</div>
`;
